spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala

```diff
@@ -44,8 +44,12 @@ class RoutingKeyGenerator(table: TableDef, columnNames: Seq[String])
 
   private def fillRoutingKey(stmt: BoundStatement): Array[ByteBuffer] = {
     val rk = routingKey.get
-    for (i <- 0 until partitionKeyIdxs.size)
+    for (i ← partitionKeyIdxs.indices) {
+      if (stmt.isNull(partitionKeyIdxs(i)))
+        throw new NullPointerException(
```

Comment:
IMO never throw an NPE, in fact, why do you do _any_ work prior to checking the argument to the function?

        private def fillRoutingKey(stmt: BoundStatement): Array[ByteBuffer] = {
           require(stmt != null, "BoundStatement 'stmt' must have a valid value.") 
        }

then in your specs check for IllegalArgumentException, because really you are trying to alert the engineer in runtime that you can't pass nulls to this api function. If this is called by our api then we'd better be checking the arg state!